File: demo/flutter/flutternew/lib/service/ali_chat_service.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import '../config/module_key_config.dart';

class AliChatService {
  Future<void> sendChat({
    required String systemPrompt,
    required List<Map<String, String>> history,
    required String userMessage,
    required void Function(String deltaText) onDelta,
    required void Function() onDone,
  }) async {
    final moduleConfig = await ModuleKeyConfigStore.load();
    final llmConfig = moduleConfig.llmAli;

    final requestBody = _buildRequestBody(
      llmConfig: llmConfig,
      systemPrompt: systemPrompt,
      history: history,
      userMessage: userMessage,
    );

    final client = HttpClient();
    bool doneNotified = false;

    void notifyDone() {
      if (doneNotified) {
        return;
      }
      doneNotified = true;
      onDone();
    }

    try {
      final request = await client.postUrl(Uri.parse(llmConfig.hostUrl));
      request.headers.set(
        HttpHeaders.authorizationHeader,
        'Bearer ${llmConfig.apiKey}',
      );
      request.headers.set(
        HttpHeaders.contentTypeHeader,
        'application/json',
      );
      request.add(utf8.encode(jsonEncode(requestBody)));

      final response = await request.close();
      if (response.statusCode < 200 || response.statusCode >= 300) {
        final errorBody = await response.transform(utf8.decoder).join();
        throw Exception(
          'Ali LLM请求失败: status=${response.statusCode}, body=$errorBody',
        );
      }

      String buffer = '';
      await for (final String chunk in response.transform(utf8.decoder)) {
        buffer += chunk;
        buffer = _consumeSseBuffer(
          buffer,
          onDelta: onDelta,
          onDone: notifyDone,
        );
      }
      final tail = buffer.trim();
      if (tail.isNotEmpty) {
        _handleSseLine(
          tail,
          onDelta: onDelta,
          onDone: notifyDone,
        );
      }

      notifyDone();
    } finally {
      client.close(force: true);
    }
  }

  Map<String, dynamic> _buildRequestBody({
    required AliLlmKeyConfig llmConfig,
    required String systemPrompt,
    required List<Map<String, String>> history,
    required String userMessage,
  }) {
    final List<Map<String, String>> messages = <Map<String, String>>[];
    if (systemPrompt.trim().isNotEmpty) {
      messages.add(<String, String>{
        'role': 'system',
        'content': systemPrompt.trim(),
      });
    }

    for (final item in history) {
      final content = (item['content'] ?? '').trim();
      if (content.isEmpty) {
        continue;
      }
      messages.add(<String, String>{
        'role': _normalizeRole(item['role']),
        'content': content,
      });
    }

    messages.add(<String, String>{
      'role': 'user',
      'content': userMessage,
    });

    return <String, dynamic>{
      'model': llmConfig.model,
      'messages': messages,
      'stream': true,
      'temperature': llmConfig.temperature,
      'max_tokens': llmConfig.maxTokens,
    };
  }

  String _normalizeRole(String? role) {
    switch ((role ?? '').toLowerCase()) {
      case 'system':
      case 'assistant':
      case 'user':
        return role!.toLowerCase();
      default:
        return 'user';
    }
  }

  String _consumeSseBuffer(
    String buffer, {
    required void Function(String deltaText) onDelta,
    required void Function() onDone,
  }) {
    while (true) {
      final int newLineIndex = buffer.indexOf('\n');
      if (newLineIndex < 0) {
        return buffer;
      }
      final String line = buffer.substring(0, newLineIndex).trim();
      buffer = buffer.substring(newLineIndex + 1);
      if (line.isEmpty) {
        continue;
      }
      _handleSseLine(line, onDelta: onDelta, onDone: onDone);
    }
  }

  void _handleSseLine(
    String line, {
    required void Function(String deltaText) onDelta,
    required void Function() onDone,
  }) {
    if (!line.startsWith('data:')) {
      return;
    }

    final String data = line.substring(5).trim();
    if (data.isEmpty) {
      return;
    }
    if (data == '[DONE]') {
      onDone();
      return;
    }

    final Map<String, dynamic> jsonMap = jsonDecode(data) as Map<String, dynamic>;
    if (jsonMap['error'] != null) {
      throw Exception('Ali LLM返回错误: ${jsonMap['error']}');
    }

    final List<dynamic> choices = jsonMap['choices'] as List<dynamic>? ?? <dynamic>[];
    if (choices.isEmpty) {
      return;
    }

    final Map<String, dynamic> firstChoice =
        (choices.first as Map?)?.cast<String, dynamic>() ?? <String, dynamic>{};
    final Map<String, dynamic> delta =
        (firstChoice['delta'] as Map?)?.cast<String, dynamic>() ?? <String, dynamic>{};
    final String deltaText = (delta['content'] ?? '').toString();
    if (deltaText.isNotEmpty) {
      onDelta(deltaText);
    }

    final String finishReason = (firstChoice['finish_reason'] ?? '').toString();
    if (finishReason.isNotEmpty && finishReason != 'null') {
      onDone();
    }
  }
}

File: demo/app/app/src/main/java/com/vectordemo/service/ai/AliChatService.kt
package com.vectordemo.service.ai

import com.vectordemo.config.ModuleKeyConfigStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject

class AliChatService(
    private val configProvider: () -> com.vectordemo.config.AliLlmKeyConfig
) : ChatService {

    private val client = OkHttpClient.Builder().build()

    override suspend fun sendChat(
        systemPrompt: String,
        history: List<Map<String, String>>,
        userMessage: String,
        onDelta: (String) -> Unit,
        onDone: () -> Unit
    ) = withContext(Dispatchers.IO) {
        val cfg = configProvider()
        val body = buildBody(cfg, systemPrompt, history, userMessage)
        val request = Request.Builder()
            .url(cfg.hostUrl)
            .addHeader("Authorization", "Bearer ${cfg.apiKey}")
            .addHeader("Content-Type", "application/json")
            .post(body.toString().toRequestBody("application/json".toMediaType()))
            .build()

        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw IllegalStateException("Ali LLM请求失败: ${response.code} ${response.body?.string().orEmpty()}")
            }
            val source = response.body?.source() ?: return@use
            var done = false
            while (!source.exhausted()) {
                val line = source.readUtf8Line()?.trim().orEmpty()
                if (!line.startsWith("data:")) {
                    continue
                }
                val payload = line.removePrefix("data:").trim()
                if (payload.isEmpty()) {
                    continue
                }
                if (payload == "[DONE]") {
                    done = true
                    onDone()
                    break
                }
                val json = JSONObject(payload)
                if (json.has("error")) {
                    throw IllegalStateException("Ali LLM返回错误: ${json.get("error")}")
                }
                val choices = json.optJSONArray("choices") ?: JSONArray()
                if (choices.length() <= 0) {
                    continue
                }
                val choice = choices.optJSONObject(0) ?: continue
                val delta = choice.optJSONObject("delta")
                val text = delta?.optString("content").orEmpty()
                if (text.isNotEmpty()) {
                    onDelta(text)
                }
                val finishReason = choice.optString("finish_reason")
                if (finishReason.isNotEmpty() && finishReason != "null") {
                    done = true
                    onDone()
                    break
                }
            }
            if (!done) {
                onDone()
            }
        }
    }

    private fun buildBody(
        cfg: com.vectordemo.config.AliLlmKeyConfig,
        systemPrompt: String,
        history: List<Map<String, String>>,
        userMessage: String
    ): JSONObject {
        val messages = JSONArray()
        if (systemPrompt.isNotBlank()) {
            messages.put(JSONObject().put("role", "system").put("content", systemPrompt.trim()))
        }
        history.forEach { item ->
            val content = item["content"].orEmpty().trim()
            if (content.isBlank()) return@forEach
            val role = item["role"].orEmpty().lowercase().let {
                if (it == "system" || it == "assistant" || it == "user") it else "user"
            }
            messages.put(JSONObject().put("role", role).put("content", content))
        }
        messages.put(JSONObject().put("role", "user").put("content", userMessage))
        return JSONObject()
            .put("model", cfg.model)
            .put("messages", messages)
            .put("stream", true)
            .put("temperature", cfg.temperature)
            .put("max_tokens", cfg.maxTokens)
    }
}

